### PR TITLE
Add nullable types detection for return values and parameters. Improve local variables scoping.

### LINKED ISSUE #1276: Extract method finds incorrect return value
In that code (extraction is surrounded by `<>`):
```php
	private function verify(string $templateName, string $email): void
	{
		<>
		if (!empty($templateName)) {
			try {
				$this->clientProvider->getSesClient()->sendCustomVerificationEmail([
				'EmailAddress' => $email,
				'TemplateName' => $templateName,
			]);
			} catch (AwsException $e) {
				throw new VerificationFailed("Could not verify.", 0, $e);
			}
		} else {
			try {
				$this->clientProvider->getSesClient()->verifyEmailIdentity([
				'EmailAddress' => $email,
			]);
			} catch (AwsException $e) {
				throw new VerificationFailed("Could not verify.", 0, $e);
			}
		}
		<>

		try {
			$this->clientProvider->getSesClient()->verifyEmailIdentity([
			'EmailAddress' => $email,
		]);
		} catch (AwsException $e) {
			throw new VerificationFailed("Could not verify.", 0, $e);
		}
	}
```
After method extraction:
```php
	private function verify(string $templateName, string $email): void
	{
		$e = $this->newMethod($templateName, $email);

		try {
			$this->clientProvider->getSesClient()->verifyEmailIdentity([
			'EmailAddress' => $email,
		]);
		} catch (AwsException $e) {
			throw new VerificationFailed("Could not verify.", 0, $e);
		}
	}

	private function newMethod(string $templateName, string $email): AwsException
	{
		if (!empty($templateName)) {
			try {
				$this->clientProvider->getSesClient()->sendCustomVerificationEmail([
				'EmailAddress' => $email,
				'TemplateName' => $templateName,
			]);
			} catch (AwsException $e) {
				throw new VerificationFailed("Could not verify.", 0, $e);
			}
		} else {
			try {
				$this->clientProvider->getSesClient()->verifyEmailIdentity([
				'EmailAddress' => $email,
			]);
			} catch (AwsException $e) {
				throw new VerificationFailed("Could not verify.", 0, $e);
			}
		}
		return $e;
	}
```

There is a bogus return value of type `AwsException`.

### LINKED ISSUE #1278: Extract method does not detect nullable types
Variables defined as `?string` are created as `string` in extracted method arguments.

### PR DESCRIPTION
**Pull request by BladeMF**
Fixes https://github.com/phpactor/phpactor/issues/1276
Fixes https://github.com/phpactor/phpactor/issues/1278.

### CODE REVIEW

File: lib/Adapter/WorseReflection/Refactor/WorseExtractMethod.php [hunk 1]
```diff
@@ -263,10 +264,10 @@ private function addParametersAndGetArgs(array $freeVariables, MethodBuilder $me
             }
 
             $parameterBuilder = $methodBuilder->parameter($freeVariable->name());
-            $variableType = $freeVariable->symbolContext()->type();
-
+            $variableType = $freeVariable->symbolContext()->types()->best();
             if ($variableType->isDefined()) {
-                $parameterBuilder->type($variableType->short());
+                $prefix = $variableType->isNullable() ? '?' : '';
```
Comment (dantleech) [inline lib/Adapter/WorseReflection/Refactor/WorseExtractMethod.php:269]: this should rather be done here I think: https://github.com/phpactor/code-builder/blob/master/templates/MethodHeader.php.twig#L14

**Comment by BladeMF** [comment]
A gentle reminder?

File: lib/Adapter/WorseReflection/Refactor/WorseExtractMethod.php [hunk 2]
```diff
@@ -289,20 +290,36 @@ private function scopeLocalVariables(SourceCode $source, int $offsetStart, int $
...
+                $text = $nodeOrToken->getText($fileContents);
+                if (is_string($text) && !in_array($text, $ignoreNames)) {
                     $name = substr($text, 1);
                     $variables[$name] = $name;
                 }
             }
         }
-
-        return array_values($variables);
+        return $variables;
```
Comment (dantleech) [inline lib/Adapter/WorseReflection/Refactor/WorseExtractMethod.php:322]: why is array_values removed?

**Comment by dantleech** [comment]
Thanks!

**Pull request merged**